omegaconf/dictconfig.py

```diff
@@ -158,17 +158,28 @@ def _validate_set(self, key: Any, value: Any) -> None:
 
         target = self._get_node(key) if key is not None else self
 
-        target_has_ref_type = isinstance(
-            target, DictConfig
-        ) and target._metadata.ref_type not in (Any, dict)
-        is_valid_target = target is None or not target_has_ref_type
+        target_type = (
+            target._metadata.ref_type
+            if target is not None
+            else self._metadata.element_type
+        )
+        target_has_ref_type = isinstance(target, DictConfig) and target_type not in (
+            Any,
+            dict,
+        )
+        value_type = OmegaConf.get_type(value)
+
+        input_container = (
+            is_structured_config(value_type) and self._metadata.element_type is not Any
+        )
```

Comment:
Oops, it should be `input_structured_config`. If value_type is a structured and we have an element_type it should do `issubclass(value_type, target_type)` which didn't happen because this case was considered a "valid_value"

Reply:
self._metadata.element_type can also be a primitive type. (int, str etc).
I think you should reorganize the logic here a bit.